# بسم الله الرحمن الرحيم

In [1]:
# import pandas as pd 
# import numpy as np 
# from matplotlib import pyplot as plt 
# import seaborn as sns 
# plt.style.use('default')
# #-------------------------------------------------------------------------------------
# from sklearn.utils import resample 
# from sklearn import preprocessing
# from sklearn.linear_model import LogisticRegression
# from sklearn.model_selection import train_test_split
# from sklearn.ensemble import RandomForestClassifier
# from sklearn.tree import DecisionTreeClassifier
# from sklearn.svm import SVC
# from sklearn.pipeline import make_pipeline
# from sklearn.metrics import confusion_matrix
# from sklearn.model_selection import cross_validate

# from sklearn.naive_bayes import GaussianNB
# from sklearn.metrics import confusion_matrix
# from sklearn.metrics import classification_report
# from sklearn.metrics import precision_score
# from sklearn.metrics import recall_score
# from sklearn.pipeline import Pipeline 

# import tensorflow_addons as tfa
# import tensorflow as tf 
# from tensorflow import keras 
# from tensorflow.keras import layers 
# from focal_loss import SparseCategoricalFocalLoss

# #-------------------------------------------------------------------------------------

# from keras.callbacks import EarlyStopping
# from keras.callbacks import ModelCheckpoint
# #-------------------------------------------------------------------------------------
# from imblearn.over_sampling import SMOTE
# from utils import *
# %matplotlib inline

In [2]:
from utils import *
import pandas as pd 
from sklearn.model_selection import train_test_split


2023-05-13 10:07:27.461972: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-05-13 10:07:27.635406: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2023-05-13 10:07:27.635433: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2023-05-13 10:07:28.497251: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory
2023-

## Reading the Dataset, cleaning , preparing it for the training 

In [3]:
df = pd.read_csv("dataset/body_level_classification_train.csv")

In [4]:
df_original= cleaning_data(df).copy()

In [5]:
all_class_train,all_class_test=data_spliting(df_original,1/3)

In [6]:
X_train=all_class_train.drop(["Body_Level"],axis=1).to_numpy()
Y_train=all_class_train["Body_Level"].to_numpy()
X_test =all_class_test.drop(["Body_Level"],axis=1).to_numpy()
Y_test =all_class_test["Body_Level"].to_numpy()

In [7]:
X_train, scaler= data_scaling(X_train)


In [8]:
X_val, X_test, Y_val, Y_test = train_test_split(X_test
                                                     , Y_test, test_size=0.5, random_state=42)

In [9]:
X_val = scaler.transform(X_val)
X_test=scaler.transform(X_test)

In [10]:
model=build_model(7,0.01)

2023-05-13 10:07:30.062653: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2023-05-13 10:07:30.062710: W tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: UNKNOWN ERROR (303)
2023-05-13 10:07:30.062760: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (brownboyComputer): /proc/driver/nvidia/version does not exist
2023-05-13 10:07:30.063341: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [11]:
history = model.fit(X_train,Y_train,validation_data=[X_val,Y_val] 
             ,batch_size=32, epochs=500 , verbose =1, 
                       callbacks=[CustomCallback("NNModels/before_sampling.h5")])

Epoch 1/500
31/31 [==============================] - 1s 9ms/step - loss: 0.4713 - accuracy: 0.3621 - val_loss: 0.2158 - val_accuracy: 0.5429
Epoch 2/500
31/31 [==============================] - 0s 2ms/step - loss: 0.1517 - accuracy: 0.5578 - val_loss: 0.1129 - val_accuracy: 0.5755
Epoch 3/500
31/31 [==============================] - 0s 2ms/step - loss: 0.0870 - accuracy: 0.6379 - val_loss: 0.0765 - val_accuracy: 0.6653
Epoch 4/500
31/31 [==============================] - 0s 2ms/step - loss: 0.0607 - accuracy: 0.6643 - val_loss: 0.0645 - val_accuracy: 0.6857
Epoch 5/500
31/31 [==============================] - 0s 3ms/step - loss: 0.0467 - accuracy: 0.6968 - val_loss: 0.0585 - val_accuracy: 0.7306
Epoch 6/500
31/31 [==============================] - 0s 3ms/step - loss: 0.0413 - accuracy: 0.7302 - val_loss: 0.0538 - val_accuracy: 0.7592
Epoch 7/500
31/31 [==============================] - 0s 3ms/step - loss: 0.0371 - accuracy: 0.7485 - val_loss: 0.0511 - val_accuracy: 0.7755
Epoch 8/500
3

Epoch 47/500
26/31 [========================>.....] - ETA: 0s - loss: 0.0074 - accuracy: 0.9075
Yes You are here 0.9137930870056152 0.9142857193946838
31/31 [==============================] - 0s 4ms/step - loss: 0.0073 - accuracy: 0.9138 - val_loss: 0.0141 - val_accuracy: 0.9143
Epoch 48/500
28/31 [==========================>...] - ETA: 0s - loss: 0.0073 - accuracy: 0.9029
Yes You are here 0.906693696975708 0.9142857193946838
31/31 [==============================] - 0s 4ms/step - loss: 0.0073 - accuracy: 0.9067 - val_loss: 0.0137 - val_accuracy: 0.9143
Epoch 49/500
30/31 [============================>.] - ETA: 0s - loss: 0.0070 - accuracy: 0.9187
Yes You are here 0.9168357253074646 0.918367326259613
31/31 [==============================] - 0s 3ms/step - loss: 0.0070 - accuracy: 0.9168 - val_loss: 0.0137 - val_accuracy: 0.9184
Epoch 50/500
31/31 [==============================] - ETA: 0s - loss: 0.0069 - accuracy: 0.9168
Yes You are here 0.9168357253074646 0.918367326259613
31/31 [=====

31/31 [==============================] - 0s 2ms/step - loss: 0.0036 - accuracy: 0.9462 - val_loss: 0.0091 - val_accuracy: 0.9633
Epoch 85/500
31/31 [==============================] - 0s 2ms/step - loss: 0.0036 - accuracy: 0.9533 - val_loss: 0.0085 - val_accuracy: 0.9510
Epoch 86/500
 1/31 [..............................] - ETA: 0s - loss: 0.0048 - accuracy: 0.9688
Yes You are here 0.9513184428215027 0.9591836929321289
31/31 [==============================] - 0s 2ms/step - loss: 0.0035 - accuracy: 0.9513 - val_loss: 0.0088 - val_accuracy: 0.9592
Epoch 87/500
31/31 [==============================] - 0s 3ms/step - loss: 0.0035 - accuracy: 0.9452 - val_loss: 0.0085 - val_accuracy: 0.9551
Epoch 88/500
31/31 [==============================] - 0s 4ms/step - loss: 0.0035 - accuracy: 0.9442 - val_loss: 0.0082 - val_accuracy: 0.9551
Epoch 89/500
31/31 [==============================] - 0s 3ms/step - loss: 0.0034 - accuracy: 0.9503 - val_loss: 0.0084 - val_accuracy: 0.9551
Epoch 90/500
31/31 [===

Epoch 129/500
31/31 [==============================] - 0s 4ms/step - loss: 0.0020 - accuracy: 0.9645 - val_loss: 0.0060 - val_accuracy: 0.9796
Epoch 130/500
 1/31 [..............................] - ETA: 0s - loss: 0.0014 - accuracy: 1.0000
Yes You are here 0.9695740342140198 0.9755101799964905
31/31 [==============================] - 0s 4ms/step - loss: 0.0019 - accuracy: 0.9696 - val_loss: 0.0061 - val_accuracy: 0.9755
Epoch 131/500
31/31 [==============================] - 0s 4ms/step - loss: 0.0020 - accuracy: 0.9604 - val_loss: 0.0067 - val_accuracy: 0.9714
Epoch 132/500
27/31 [=========================>....] - ETA: 0s - loss: 0.0019 - accuracy: 0.9711
Yes You are here 0.970588207244873 0.9755101799964905
31/31 [==============================] - 0s 4ms/step - loss: 0.0019 - accuracy: 0.9706 - val_loss: 0.0064 - val_accuracy: 0.9755
Epoch 133/500
31/31 [==============================] - 0s 4ms/step - loss: 0.0019 - accuracy: 0.9655 - val_loss: 0.0066 - val_accuracy: 0.9796
Epoch 134/

 1/31 [..............................] - ETA: 0s - loss: 5.7445e-04 - accuracy: 1.0000
Yes You are here 0.9817444086074829 0.9836734533309937
31/31 [==============================] - 0s 3ms/step - loss: 0.0011 - accuracy: 0.9817 - val_loss: 0.0050 - val_accuracy: 0.9837
Epoch 176/500
31/31 [==============================] - 0s 2ms/step - loss: 0.0011 - accuracy: 0.9828 - val_loss: 0.0053 - val_accuracy: 0.9796
Epoch 177/500
31/31 [==============================] - 0s 3ms/step - loss: 0.0011 - accuracy: 0.9787 - val_loss: 0.0050 - val_accuracy: 0.9796
Epoch 178/500
31/31 [==============================] - 0s 2ms/step - loss: 0.0011 - accuracy: 0.9807 - val_loss: 0.0051 - val_accuracy: 0.9796
Epoch 179/500
31/31 [==============================] - 0s 3ms/step - loss: 0.0011 - accuracy: 0.9797 - val_loss: 0.0049 - val_accuracy: 0.9796
Epoch 180/500
31/31 [==============================] - 0s 3ms/step - loss: 0.0011 - accuracy: 0.9807 - val_loss: 0.0050 - val_accuracy: 0.9796
Epoch 181/500


Epoch 221/500
31/31 [==============================] - 0s 3ms/step - loss: 8.5193e-04 - accuracy: 0.9797 - val_loss: 0.0054 - val_accuracy: 0.9796
Epoch 222/500
31/31 [==============================] - 0s 3ms/step - loss: 8.2536e-04 - accuracy: 0.9838 - val_loss: 0.0050 - val_accuracy: 0.9837
Epoch 223/500
31/31 [==============================] - 0s 3ms/step - loss: 8.4521e-04 - accuracy: 0.9848 - val_loss: 0.0052 - val_accuracy: 0.9837
Epoch 224/500
31/31 [==============================] - 0s 3ms/step - loss: 0.0010 - accuracy: 0.9716 - val_loss: 0.0054 - val_accuracy: 0.9796
Epoch 225/500
31/31 [==============================] - 0s 3ms/step - loss: 8.9249e-04 - accuracy: 0.9807 - val_loss: 0.0046 - val_accuracy: 0.9837
Epoch 226/500
31/31 [==============================] - 0s 4ms/step - loss: 9.1858e-04 - accuracy: 0.9767 - val_loss: 0.0042 - val_accuracy: 0.9673
Epoch 227/500
31/31 [==============================] - 0s 4ms/step - loss: 9.6875e-04 - accuracy: 0.9767 - val_loss: 0.004

31/31 [==============================] - 0s 2ms/step - loss: 5.1211e-04 - accuracy: 0.9888 - val_loss: 0.0033 - val_accuracy: 0.9837
Epoch 274/500
31/31 [==============================] - 0s 2ms/step - loss: 5.2016e-04 - accuracy: 0.9868 - val_loss: 0.0034 - val_accuracy: 0.9837
Epoch 275/500
31/31 [==============================] - 0s 2ms/step - loss: 4.9858e-04 - accuracy: 0.9888 - val_loss: 0.0032 - val_accuracy: 0.9837
Epoch 276/500
31/31 [==============================] - 0s 3ms/step - loss: 5.0146e-04 - accuracy: 0.9888 - val_loss: 0.0032 - val_accuracy: 0.9837
Epoch 277/500
31/31 [==============================] - 0s 4ms/step - loss: 4.8839e-04 - accuracy: 0.9919 - val_loss: 0.0033 - val_accuracy: 0.9837
Epoch 278/500
31/31 [==============================] - 0s 4ms/step - loss: 4.9588e-04 - accuracy: 0.9909 - val_loss: 0.0032 - val_accuracy: 0.9837
Epoch 279/500
31/31 [==============================] - 0s 4ms/step - loss: 4.8815e-04 - accuracy: 0.9909 - val_loss: 0.0032 - val_ac

Epoch 328/500
31/31 [==============================] - 0s 2ms/step - loss: 4.3748e-04 - accuracy: 0.9899 - val_loss: 0.0039 - val_accuracy: 0.9796
Epoch 329/500
31/31 [==============================] - 0s 2ms/step - loss: 4.6822e-04 - accuracy: 0.9858 - val_loss: 0.0054 - val_accuracy: 0.9796
Epoch 330/500
31/31 [==============================] - 0s 2ms/step - loss: 4.5087e-04 - accuracy: 0.9888 - val_loss: 0.0051 - val_accuracy: 0.9796
Epoch 331/500
31/31 [==============================] - 0s 2ms/step - loss: 4.3550e-04 - accuracy: 0.9888 - val_loss: 0.0049 - val_accuracy: 0.9837
Epoch 332/500
31/31 [==============================] - 0s 2ms/step - loss: 4.6209e-04 - accuracy: 0.9878 - val_loss: 0.0047 - val_accuracy: 0.9837
Epoch 333/500
31/31 [==============================] - 0s 2ms/step - loss: 4.0541e-04 - accuracy: 0.9868 - val_loss: 0.0047 - val_accuracy: 0.9837
Epoch 334/500
31/31 [==============================] - 0s 2ms/step - loss: 4.1563e-04 - accuracy: 0.9868 - val_loss: 0

Epoch 384/500
31/31 [==============================] - 0s 2ms/step - loss: 3.3925e-04 - accuracy: 0.9919 - val_loss: 0.0064 - val_accuracy: 0.9837
Epoch 385/500
31/31 [==============================] - 0s 2ms/step - loss: 3.2857e-04 - accuracy: 0.9909 - val_loss: 0.0068 - val_accuracy: 0.9796
Epoch 386/500
31/31 [==============================] - 0s 2ms/step - loss: 3.2761e-04 - accuracy: 0.9909 - val_loss: 0.0064 - val_accuracy: 0.9796
Epoch 387/500
31/31 [==============================] - 0s 2ms/step - loss: 3.2594e-04 - accuracy: 0.9929 - val_loss: 0.0068 - val_accuracy: 0.9796
Epoch 388/500
31/31 [==============================] - 0s 2ms/step - loss: 3.2124e-04 - accuracy: 0.9919 - val_loss: 0.0066 - val_accuracy: 0.9837
Epoch 389/500
31/31 [==============================] - 0s 2ms/step - loss: 3.3500e-04 - accuracy: 0.9919 - val_loss: 0.0066 - val_accuracy: 0.9796
Epoch 390/500
31/31 [==============================] - 0s 2ms/step - loss: 3.2617e-04 - accuracy: 0.9919 - val_loss: 0

Epoch 440/500
31/31 [==============================] - 0s 2ms/step - loss: 0.0018 - accuracy: 0.9615 - val_loss: 0.0063 - val_accuracy: 0.9714
Epoch 441/500
31/31 [==============================] - 0s 2ms/step - loss: 0.0033 - accuracy: 0.9371 - val_loss: 0.0125 - val_accuracy: 0.9633
Epoch 442/500
31/31 [==============================] - 0s 2ms/step - loss: 0.0046 - accuracy: 0.9462 - val_loss: 0.0250 - val_accuracy: 0.9061
Epoch 443/500
31/31 [==============================] - 0s 2ms/step - loss: 0.0064 - accuracy: 0.9473 - val_loss: 0.0146 - val_accuracy: 0.9347
Epoch 444/500
31/31 [==============================] - 0s 2ms/step - loss: 0.0024 - accuracy: 0.9462 - val_loss: 0.0109 - val_accuracy: 0.9592
Epoch 445/500
31/31 [==============================] - 0s 2ms/step - loss: 0.0024 - accuracy: 0.9564 - val_loss: 0.0107 - val_accuracy: 0.9633
Epoch 446/500
31/31 [==============================] - 0s 2ms/step - loss: 0.0014 - accuracy: 0.9716 - val_loss: 0.0111 - val_accuracy: 0.9429

Epoch 496/500
31/31 [==============================] - 0s 3ms/step - loss: 2.7822e-04 - accuracy: 0.9919 - val_loss: 0.0081 - val_accuracy: 0.9796
Epoch 497/500
31/31 [==============================] - 0s 3ms/step - loss: 2.5816e-04 - accuracy: 0.9919 - val_loss: 0.0085 - val_accuracy: 0.9796
Epoch 498/500
31/31 [==============================] - 0s 2ms/step - loss: 2.8029e-04 - accuracy: 0.9878 - val_loss: 0.0087 - val_accuracy: 0.9796
Epoch 499/500
31/31 [==============================] - 0s 2ms/step - loss: 2.6655e-04 - accuracy: 0.9929 - val_loss: 0.0084 - val_accuracy: 0.9796
Epoch 500/500
31/31 [==============================] - 0s 2ms/step - loss: 2.8952e-04 - accuracy: 0.9919 - val_loss: 0.0091 - val_accuracy: 0.9755


In [13]:
accurate_train_accuracy,classify_report= test_model(model,X_train,Y_train)

31/31 [==============================] - 0s 901us/step


In [14]:
accurate_train_accuracy

0.9949290060851927

In [15]:
print(classify_report)

              precision    recall  f1-score   support

           0       0.98      1.00      0.99       127
           1       0.98      0.98      0.98       134
           2       1.00      0.99      0.99       271
           3       1.00      1.00      1.00       454

    accuracy                           0.99       986
   macro avg       0.99      0.99      0.99       986
weighted avg       0.99      0.99      0.99       986



In [16]:
accurate_val_accuracy,classify_report= test_model(model,X_val,Y_val)

8/8 [==============================] - 0s 1ms/step


In [17]:
accurate_val_accuracy

0.9755102040816327

In [18]:
print(classify_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        29
           1       0.97      0.97      0.97        31
           2       0.95      0.97      0.96        73
           3       0.99      0.97      0.98       112

    accuracy                           0.98       245
   macro avg       0.98      0.98      0.98       245
weighted avg       0.98      0.98      0.98       245



In [19]:
accurate_test_accuracy,classify_report= test_model(model,X_test,Y_test)

8/8 [==============================] - 0s 1ms/step


In [20]:
accurate_test_accuracy

0.975609756097561

In [ ]:
print(classify_report)

## important-Note
gamma=1 on-test(0.967479674796748) on-val(0.9836734693877551)
gamma=2 on-test(0.967479674796748) on-val(0.9836734693877551) 
gamma=3 on-test(0.975609756097561) on-val(0.9836734693877551)